### ISSUE TITLE
How do you get the Play server to stop? Ctrl-C only stops React, it seems...

### ISSUE THREAD

**Issue by toddobryan** [yohangz/scala-play-react-seed#20]
When I `sbt run` in the terminal and then `Ctrl-C` to quit, I get
```
[warn] Canceling execution...
```
but it never finishes and the Play server keeps running. I can't re-run things because I get a BindException since port 9000 is still being used.

I'm doing this in the IntelliJ sbt shell, and quitting the shell doesn't work. I have to completely quit IntelliJ to get the process to release the port.

Running in a separate terminal, the same thing happens. Supposedly, you can exit out of the Play server by pressing enter, but that doesn't seem to work at all. At least with the terminal, if I exit out of the terminal then the Play server releases the port so I can `sbt run` again.

Anyone else having this problem or am I just doing something silly?

**Comment by yohangz** [comment]
@toddobryan, can you try the same with a fresh project (without using this seed) and share the observations. I believe it's something to do with your environment.

**Comment by perNyfelt** [comment]
I have a similar (maybe exactly the same) issue. In the sbt console, executing run works fine (node on 3000 and Play on port 9000 etc.). However when doing ctrl + d to stop the server, node does not exit. When trying to start again i get "Something is already running on port 3000." From that point on whatever changes I do to react components are not shown anymore. After finding the process and killing it, starting again works fine. 

adding "stop": "pkill --signal SIGINT node" to package.json scripts section allows me to stop it by doing "npm stop" in a terminal but I have not found a way to get that to be called automatically by adding a call to it afterStopped() in the FrontendRunHook when I stop Play. Any suggestions?

**Comment by yohangz** [comment]
@perNyfelt, can you please confirm the target OS? I assume it's Linux. I don't recommend killing the node process. However, we can kill the process listening to port 3000. This solution should be the last resort. Let's try to resolve it within play run hook. I've made a slight change in play run hook, Please confirm whether there are any changes. I will try to recreate this in Linux.

**Comment by perNyfelt** [comment]
Yes, I am running Linux Mint 19.3, Play 2.7.4, Scala 2.13.1 and sbt 1.3.6. Unfortunately the fix you did does not solve the problem for me. I still get 

`Something is already running on port 3000.`

When i shut down using ctrl+d and then start again with run.

I tried with kill-port 3000 as the stop command instead but although it works it gives med the following error when shutting down:
```
Process on port 3000 killed
Exception in thread "Thread-170" java.io.IOException: Stream closed
	at java.io.BufferedInputStream.getBufIfOpen(BufferedInputStream.java:170)
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:283)
	at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
	at java.io.FilterInputStream.read(FilterInputStream.java:107)
	at scala.sys.process.BasicIO$.loop$1(BasicIO.scala:238)
	at scala.sys.process.BasicIO$.transferFullyImpl(BasicIO.scala:246)
	at scala.sys.process.BasicIO$.transferFully(BasicIO.scala:227)
	at scala.sys.process.BasicIO$.$anonfun$toStdOut$1(BasicIO.scala:221)
	at scala.sys.process.BasicIO$.$anonfun$toStdOut$1$adapted(BasicIO.scala:221)
	at scala.sys.process.ProcessBuilderImpl$Simple.$anonfun$run$3(ProcessBuilderImpl.scala:80)
	at scala.sys.process.ProcessImpl$Spawn$$anon$1.run(ProcessImpl.scala:27)
```
However, changing the stop command to `kill $(lsof -t -i:3000)` is probably a better idea and works just as fine as `pkill --signal SIGINT node` did.

**Comment by yohangz** [comment]
@perNyfelt, I'm still thinking of a way to resolve this issue without killing the process explicitly (as we already have code in place for that purpose). I'll keep you posted.

**Comment by cagline** [comment]
@toddobryan , @perNyfelt 
Try to recreate on this issue on my machine by runing `sbt run` and then `Ctrl-C` (same e has been executed 4 times),  It worked smoothly on terminal & intelliJ terminal without any issue.

```
Operating System: Ubuntu 16.04.4 LTS
Kernel: Linux 4.15.0-72-generic
Architecture: x86-64
```

**Comment by perNyfelt** [comment]
How about when doing ctrl+d instead of ctrl+c?

**Comment by toddobryan** [comment]
Feel free to close this. Given that we have to start play and react separately with react-scripts >= 3.0, this isn't so much of an issue.

**Comment by perNyfelt** [comment]
OK, makes sense

**Comment by AlexTo** [comment]
Seems like adding the fix in #25  can make both Play and React able to start 
```
override def afterStarted(): Unit = {
        process = Some(
          Process(run, base / "ui", "CI" -> "true").run
        )
      }
```

However, when I stop debugging Play (in IntelliJ), React won't stop. Port 3000 is in used the next time I start debugging Play ;)

**Comment by toddobryan** [comment]
Maybe open a new issue, but it sounds like the problem is that the React server is started in a separate process, so Ctrl-C-ing the play process doesn't stop it. But it also sounds like you can't *not* start the React server in a separate process, so maybe it's a Catch-22.
